flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/AffinityPropagation.java

```diff
@@ -0,0 +1,535 @@
+
+		// Create E and I AP vertices
+		DataSet<Vertex<Long, APVertexValue>> verticesWithAllInNeighbors =
+			input.groupReduceOnEdges(new InitAPVertex(), EdgeDirection.IN);
+
+		List<Vertex<Long, APVertexValue>> APvertices = verticesWithAllInNeighbors.collect();
+
+		// Create E and I AP edges. Could this be done with some gelly functionality?
+		List<Edge<Long, NullValue>> APedges = new ArrayList<>();
+
+		for(int i = 1; i < input.numberOfVertices() + 1; i++){
+			for(int j = 1; j < input.numberOfVertices() + 1; j++){
+				APedges.add(new Edge<>(i * 10L, j * 10L + 1, NullValue.getInstance()));
+			}
+		}
+
+		DataSet<Edge<Long, NullValue>> APEdgesDS = input.getContext().fromCollection(APedges);
+		DataSet<Vertex<Long, APVertexValue>> APVerticesDS = input.getContext().fromCollection(APvertices);
+
+		ScatterGatherConfiguration parameters = new ScatterGatherConfiguration();
+		parameters.registerAggregator("convergedAggregator", new LongSumAggregator());
+
+		Graph<Long, APVertexValue, NullValue> APgraph
+			= Graph.fromDataSet(APVerticesDS, APEdgesDS, input.getContext());
+
+		return APgraph.getUndirected().runScatterGatherIteration(new APVertexUpdater(input.numberOfVertices() * 2),
+			new APMessenger(),this.maxIterations,parameters).getVertices().filter(new FilterFunction<Vertex<Long, APVertexValue>>() {
+			@Override
+			public boolean filter(Vertex<Long, APVertexValue> vertex) throws Exception {
+				return vertex.getId()%2 == 0;
+			}
+		}).map(new MapFunction<Vertex<Long, APVertexValue>, Tuple2<Long, Long>>() {
+			@Override
+			public Tuple2<Long, Long> map(Vertex<Long, APVertexValue> value) throws Exception {
+				Tuple2<Long, Long> returnValue = new Tuple2<>(value.getId()/10, value.getValue().getExemplar()/10);
+				return returnValue;
+			}
+		});
+
+	}
+
+	/**
+	* Foreach input point we have to create a pair of E,I vertices. Same structure is used for both vertex type, to
+	* diferenciate E and I vertices is used the id. Foreach input point we will create:
+	*
+	* - One E vertex with the id as the original input id * 10 + 1
+	* - One I vertex with the id as the original input id * 10
+	*
+	* This way even ids are from E type vertices and odd ids are from I vertices.
+	*
+	* It also calculates adds the weights to the I vertices. Notice that the S vertices are not created and the weights
+	* are added to the I vertices, simulating the S vertex.
+	*/
+
+	@SuppressWarnings("serial")
+	private static final class InitAPVertex implements EdgesFunction<Long, Double, Vertex<Long, APVertexValue>> {
+
+		@Override
+		public void iterateEdges(Iterable<Tuple2<Long, Edge<Long, Double>>> edges,
+								Collector<Vertex<Long, APVertexValue>> out) throws Exception {
+
+			Vertex<Long, APVertexValue> APvertexI = new Vertex<>();
+			Vertex<Long, APVertexValue> APvertexE = new Vertex<>();
+
+			Iterator<Tuple2<Long, Edge<Long, Double>>> itr = edges.iterator();
+			Tuple2<Long, Edge<Long, Double>> edge = itr.next();
```

Comment:
Nops the way is implemented could be possible to call the init function without edges. 
I'm thinking on changing the input data to be a matrix of similarities, being each of the columns/rows the points to cluster.

Thoughts?